Moves affected by Terrain: Camouflage, Secret Power & Terrain Pulse

AsparagusEduardo opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
- Camouflage's type affected by terrain.
- Secret Power's effect and animation affected by terrain. Addresses #1662 .
- Added new battle terrain defines for Secret Power and Camouflage's effects and types.
- Implemented Terrain Pulse. Addresses #1094 .
  - Without terrain:
  ![normal](https://user-images.githubusercontent.com/2904965/140649235-7d56e5a2-6848-4176-b86a-ef860462c9f2.gif)
  - Grassy Terrain:
  ![grassy](https://user-images.githubusercontent.com/2904965/140649311-1248e950-ba7b-40c3-bd8e-40bf888dcc80.gif)
  - Misty Terrain:
  ![misty](https://user-images.githubusercontent.com/2904965/140649312-65b54efd-ab9e-4dcd-98ea-3c1aa4c8f7cb.gif)
  - Psychic Terrain:
  ![psychic](https://user-images.githubusercontent.com/2904965/140649313-8655acf3-cea4-47f4-82e0-901f90e4faf3.gif)
  - Electric Terrain
  ![electric](https://user-images.githubusercontent.com/2904965/140649314-6cea0413-ce86-462c-97d2-2151f6edb963.gif)

(Originally by Fontbane, resubmitted and modified with permission).

## **Discord contact info**
AsparagusEduardo#6051

AsparagusEduardo commented:
I decided to leave the animation for later. Ready for review.

ghoulslash commented on data/battle_anim_scripts.s:
```diff
@@ -13979,7 +13979,7 @@ Move_RISING_VOLTAGE::
 
 Move_TERRAIN_PULSE::
-	end @to do:
+	goto Move_DRAGON_PULSE @Temporary, needs to setup different animations based on terrain
```
What does the animation look like in game? A simple pulse animation that changes particle color based on terrain would be relatively easy to implement

ghoulslash commented on data/battle_anim_scripts.s:
```diff
@@ -23251,16 +23251,21 @@ Move_SKY_UPPERCUT:
 	end
 
 Move_SECRET_POWER:
```
should these animations be updated to [Gen 7](https://bulbapedia.bulbagarden.net/wiki/Secret_Power_%28move%29#Generation_VII)?

AsparagusEduardo commented:
Ready for re-review
- Added animations to Terrain Pulse:
  - Without terrain:
  ![normal](https://user-images.githubusercontent.com/2904965/140649235-7d56e5a2-6848-4176-b86a-ef860462c9f2.gif)
  - Grassy and Misty Terrains:
  ![grassy](https://user-images.githubusercontent.com/2904965/140649311-1248e950-ba7b-40c3-bd8e-40bf888dcc80.gif) ![misty](https://user-images.githubusercontent.com/2904965/140649312-65b54efd-ab9e-4dcd-98ea-3c1aa4c8f7cb.gif)
  - Psychic and Terrains:
  ![psychic](https://user-images.githubusercontent.com/2904965/140649313-8655acf3-cea4-47f4-82e0-901f90e4faf3.gif) ![electric](https://user-images.githubusercontent.com/2904965/140649314-6cea0413-ce86-462c-97d2-2151f6edb963.gif)
  - Added config for Secret Power's effects and animations.
  - Added new battle terrains to complete Secret Power's, Nature Power's and Camouflage's data (plus added failsafes in case new terrains are added and are not accounted for these moves.
